backend-go/lib.go

```diff
@@ -0,0 +1,229 @@
+	CustomNcJson *nats.EncodedConn
+}
+
+// Option takes the startup options from [StartService] and modifies the startup behaviour.
+type Option func(*Options) error
+
+// WithoutNats disables the NATS initialization step in [StartService].
+func WithoutNats() Option {
+	return func(options *Options) error {
+		options.Nats = false
+		return nil
+	}
+}
+
+// WithOverwriteArgs passes additional configuration parameters to [StartService].
+// These parameters have precedence over any other configuration source.
+func WithOverwriteArgs(args map[string]string) Option {
+	return func(options *Options) error {
+		options.OverwriteArgs = args
+		return nil
+	}
+}
+
+// WithCustomNc gives [StartService] externally managed NATS connections.
+// This option prevents [StartService] to initialize another NATS connection.
+// Both arguments must be a valid NATS connection.
+func WithCustomNc(nc *nats.Conn, ncJson *nats.EncodedConn) Option {
+	return func(options *Options) error {
+		if nc == nil {
+			return errors.New("custom nats connection is not defined (nc == nil)")
+		}
+		if ncJson == nil {
+			return errors.New("custom encoded nats connection is not defined (ncJson == nil)")
+		}
+
+		options.CustomNc = nc
+		options.CustomNcJson = ncJson
+		// enable usage of NATS to register health check
+		options.Nats = true
+		return nil
+	}
+}
+
+// Service is a Telestion service that provides the available APIs.
+type Service struct {
+
+	// Nc is the NATS connection the service uses to communicate with other services on the NATS network.
+	Nc *nats.Conn
+
+	// NcJson is an abstraction of [Service.Nc]
+	// that provides automatic JSON marshaling and unmarshalling of NATS message bodies.
+	NcJson *nats.EncodedConn
+
+	// DataDir is an absolute path to the data directory of the service.
+	// This is where the service should store any data it needs to persist.
+	// The data is shared between multiple services.
+	// To ensure that the service doesn't overwrite data from other services,
+	// you should create a subdirectory for your service.
+	DataDir string
+
+	// ServiceName is the name of the service.
+	// This is used to create a unique queue group for the service.
+	ServiceName string
+
+	// Config is the assembled configuration from all available configuration sources.
+	// It configures the service.
+	Config *Config
+}
+
+// StartService starts a new Telestion service and returns the available APIs.
+// Additional options passed are applied in the order in they appear and modify the startup procedure.
+//
+// During startup the service parses the service [Config] from different configuration sources in the following order:
+//
+//  1. `overwriteArgs` from [WithOverwriteArgs]
+//  2. command line arguments
+//  3. environment variables
+//  4. default configuration, if `--dev` is passed in the steps from above
+//  5. configuration file, if `CONFIG_FILE` parameter is defined, readable and parsable
+func StartService(opts ...Option) (*Service, error) {
```

Comment:
Make the order of precedence clearer. Before, it wasn't clear, if they got applied in that order (overwriting each other) or were a list of priorities.
```suggestion
// StartService starts a new Telestion service and returns the available APIs.
// Additional options passed are applied in the order in they appear and modify the startup procedure.
//
// During startup the service parses the service [Config] from different configuration sources with the following priorities:
//
//  1. `overwriteArgs` from [WithOverwriteArgs]
//  2. command line arguments
//  3. environment variables
//  4. default configuration, if `--dev` is passed in the steps from above
//  5. configuration file, if `CONFIG_FILE` parameter is defined, readable and parsable
func StartService(opts ...Option) (*Service, error) {
```